### PR TITLE
(feat) O3-2467:  Service Queue - Back to scheduled visits

### PR DESCRIPTION
**Pull request by usamaidrsk**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary

This PR adds back to scheduled visit actions and back to search results given the flow of the user 

> Workflow 1:  When the patient has scheduled appointments within +/- 7 days of the current date
> 
> Clicking on a patient card in the patient search renders a scheduled-visits mode, in which the user can choose one of the patient’s scheduled appointments, or can click on “Start another visit type”.
> 
>  If they click this, then the mode changes to a visit_form mode in which the user completes full details of the visit to start
> 
> From here, if they want to go back to the list of appointments, they can click on the “Back to Scheduled Visits” arrow at the top of the form
> 
> Workflow 2:  When the patient has no scheduled appointments as above
> 
> Clicking on a patient card in the patient search goes straight to the visit_form mode.  In this case, there are no “Scheduled Visits” to go back to, but the option allowed for in 3 above exists.  That is the issue being reported in this ticket.
> 
> Solution:
> 
> Per the [mockups in Zeppelin](https://app.zeplin.io/project/60d59321e8100b0324762e05/screen/61bb88684d99be6bd9e2cee3), if the previous state the user was in was the patient search results, then the back arrow at the top of the form should read “Back to search results”.
> 
> If the patient is in the scheduled-visits mode via Workflow 1 or if the patient is in the visit_form mode via Workflow 2, then there should be an arrow at the top of the form to “Return to search results” which returns the user to the list of patients that they had previously selected from.

 
## Screenshots

https://github.com/openmrs/openmrs-esm-patient-management/assets/53287480/7519683a-e083-4e42-a381-b65b96687d28

## Other

[O3-2467](https://openmrs.atlassian.net/browse/O3-2467)

[O3-2467]: https://openmrs.atlassian.net/browse/O3-2467?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/esm-patient-search-app/src/patient-search-button/patient-search-button.component.tsx [hunk 1]
```diff
@@ -9,17 +9,27 @@ interface PatientSearchButtonProps {
...
   overlayHeader,
   selectPatientAction,
+  searchQueryUpdatedAction,
   buttonProps,
+  isOpen = false,
+  searchQuery = '',
 }) => {
   const { t } = useTranslation();
-  const [showSearchOverlay, setShowSearchOverlay] = useState<boolean>(false);
+  const [showSearchOverlay, setShowSearchOverlay] = useState<boolean>(isOpen);
```
Comment (chibongho) [inline packages/esm-patient-search-app/src/patient-search-button/patient-search-button.component.tsx:28]: This looks like a duplicated state. Maybe pass a `setIsOpen` function in as prop instead? (Similar to `searchQueryUpdateAction`)

File: packages/esm-service-queues-app/src/queue-table/default-queue-table.component.tsx [hunk 2]
```diff
@@ -79,15 +97,24 @@ function DefaultQueueTable() {
           <ExtensionSlot
             name="patient-search-button-slot"
             state={{
+              isOpen: isPatientSearchOpen,
```
Comment (chibongho) [inline packages/esm-service-queues-app/src/queue-table/default-queue-table.component.tsx:100]: I think we need to make similar changes to other places where the `patient-search-button-slot` is used:

esm-appointments-app/src/metrics/metrics-header.component.tsx
esm-service-queues-app/src/views/queue-tables-for-all-statuses.component.tsx

File: packages/esm-service-queues-app/src/patient-search/hooks/useScheduledVisits.ts [hunk 3]
```diff
@@ -22,7 +22,7 @@ export function useScheduledVisits(patientUuid: string) {
     });
 
   const { data, error, isLoading } = useSWR<AppointmentsFetchResponse, Error>(
-    patientUuid ? appointmentsSearchUrl : null,
+    patientUuid ? [appointmentsSearchUrl, patientUuid] : null,
```
Comment (chibongho) [inline packages/esm-service-queues-app/src/patient-search/hooks/useScheduledVisits.ts:25]: Nice fix! I didn't know you can pass multiple keys into `useSWR`. 

https://swr.vercel.app/docs/arguments

A bit minor, but maybe refactor `fetcher` to take in `appointmentsSearchUrl` and `patientUuid` params to make it more clear?

File: packages/esm-service-queues-app/src/patient-search/patient-search.workspace.tsx [hunk 4]
```diff
@@ -1,40 +1,77 @@
...
+const PatientSearch: React.FC<PatientSearchProps> = ({ closeWorkspace, viewState, handleBackToSearchList }) => {
+  const { t } = useTranslation();
   const { selectedPatientUuid } = viewState;
   const { patient } = usePatient(selectedPatientUuid);
   const { activeVisit } = useVisit(selectedPatientUuid);
   const [searchType, setSearchType] = useState<SearchTypes>(SearchTypes.SCHEDULED_VISITS);
-  const [newVisitMode, setNewVisitMode] = useState<boolean>(false);
   const [showContactDetails, setContactDetails] = useState(false);
+  const { appointments, isLoading, isError } = useScheduledVisits(selectedPatientUuid);
+  const hasAppointments = useMemo(
```
Comment (chibongho) [inline packages/esm-service-queues-app/src/patient-search/patient-search.workspace.tsx:39]: This is probably fine, but I wouldn't use `useMemo` and `useCallback` if the operations are not expensive.

https://kentcdodds.com/blog/usememo-and-usecallback#why-is-usecallback-worse

**Review by chibongho — COMMENTED** [review]
Nice work! Some feedback below.

**Review by mseaton — COMMENTED** [review]
Thanks @usamaidrsk !  I defer to @chibongho and others like @denniskigen and @brandones on the code, but the screenshots attached look great and appear to meet the requirements, nice work!

File: packages/esm-service-queues-app/src/queue-table/default-queue-table.component.tsx (hunk 2 above)
Comment (usamaidrsk) [reply packages/esm-service-queues-app/src/queue-table/default-queue-table.component.tsx:100]: The `isOpen` is an optional prop, it is only used where needed.

File: packages/esm-service-queues-app/src/patient-search/hooks/useScheduledVisits.ts (hunk 3 above)
Comment (usamaidrsk) [reply packages/esm-service-queues-app/src/patient-search/hooks/useScheduledVisits.ts:25]: Yeah @chibongho , that makes it more clear, however `fetcher` is a variable in the same scope where `appointmentsSearchUrl` and `patientUuid` are declared, thus being directly used as 
```typescript
const fetcher = () =>
    openmrsFetch(appointmentsSearchUrl, {
      method: 'POST',
      signal: abortController.signal,
      headers: {
        'Content-Type': 'application/json',
      },
      body: {
        patientUuid: patientUuid,
        startDate: startDate,
      },
    });
```

Comment (brandones) [reply packages/esm-service-queues-app/src/patient-search/hooks/useScheduledVisits.ts:25]: @usamaidrsk , a `fetcher` function is supposed to take the key as its argument, so you should write it as such. It's SWR's API so you need to follow their rules.

**Comment by brandones** [comment]
@jwnasambu reports that this also fixes https://openmrs.atlassian.net/browse/O3-3356 .

File: packages/esm-service-queues-app/src/patient-search/patient-search.workspace.tsx (hunk 4 above)
Comment (denniskigen) [reply packages/esm-service-queues-app/src/patient-search/patient-search.workspace.tsx:39]: This is a problem that will be solved soon, hopefully, with the new [React compiler](https://react.dev/learn/react-compiler#what-does-the-compiler-do).

File: packages/esm-patient-search-app/src/patient-search-button/patient-search-button.component.tsx (hunk 1 above)
Comment (brandones) [reply packages/esm-patient-search-app/src/patient-search-button/patient-search-button.component.tsx:28]: https://openmrs.atlassian.net/browse/O3-2736 :)

Comment (brandones) [reply packages/esm-patient-search-app/src/patient-search-button/patient-search-button.component.tsx:28]: Actually a lot of this code would be unnecessary with [O3-3326](https://openmrs.atlassian.net/browse/O3-3326) Migrate patient-search to use core workspace system for overlays. If this was a workspace, then the patient-search workspace would stay open in the background after a patient is selected. Then the back button would just close the present workspace, automatically re-opening the hidden patient search workspace.

Comment (brandones) [reply packages/esm-patient-search-app/src/patient-search-button/patient-search-button.component.tsx:28]: Barring that larger project, I think @usamaidrsk 's implementation is fine.

**Review by chibongho — APPROVED** [review]
LGTM!

**Pull request merged**